Velocity does not respect system/java HTTP proxy settings

I use the following command line to start volecity:

`java -DsocksProxyHost=127.0.0.1 -DsocksProxyPort=10808 -Dhttps.proxyHost=127.0.0.1 -Dhttps.proxyPort=10809 -Dhttp.proxyHost=127.0.0.1 -Dhttp.proxyPort=10809 -Djava.net.useSystemProxies=true -jar velocity-3.3.0-SNAPSHOT-320.jar`

But I found it didn't work.
However, it worked when I used these parameters to start my folia server:

`java -DsocksProxyHost=127.0.0.1 -DsocksProxyPort=10808 -Dhttps.proxyHost=127.0.0.1 -Dhttps.proxyPort=10809 -Dhttp.proxyHost=127.0.0.1 -Dhttp.proxyPort=10809 -Djava.net.useSystemProxies=true -jar server.jar nogui`

```
[16:35:45 ERROR]: Unable to authenticate with Mojang
java.util.concurrent.ExecutionException: java.net.ConnectException: connection timed out: sessionserver.mojang.com/13.107.246.74:443
        at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396) ~[?:?]
        at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073) ~[?:?]
        at org.asynchttpclient.netty.NettyResponseFuture.get(NettyResponseFuture.java:201) ~[velocity-3.3.0-SNAPSHOT-320.jar:3.3.0-SNAPSHOT (git-08c03aae-b320)]
        at com.velocitypowered.proxy.connection.client.InitialLoginSessionHandler.lambda$handle$4(InitialLoginSessionHandler.java:233) ~[velocity-3.3.0-SNAPSHOT-320.jar:3.3.0-SNAPSHOT (git-08c03aae-b320)]
        at org.asynchttpclient.netty.NettyResponseFuture.lambda$addListener$0(NettyResponseFuture.java:294) ~[velocity-3.3.0-SNAPSHOT-320.jar:3.3.0-SNAPSHOT (git-08c03aae-b320)]
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863) ~[?:?]
        at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841) ~[?:?]
        at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:482) ~[?:?]
        at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173) ~[velocity-3.3.0-SNAPSHOT-320.jar:3.3.0-SNAPSHOT (git-08c03aae-b320)]
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166) ~[velocity-3.3.0-SNAPSHOT-320.jar:3.3.0-SNAPSHOT (git-08c03aae-b320)]
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470) ~[velocity-3.3.0-SNAPSHOT-320.jar:3.3.0-SNAPSHOT (git-08c03aae-b320)]
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:569) ~[velocity-3.3.0-SNAPSHOT-320.jar:3.3.0-SNAPSHOT (git-08c03aae-b320)]
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997) ~[velocity-3.3.0-SNAPSHOT-320.jar:3.3.0-SNAPSHOT (git-08c03aae-b320)]
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[velocity-3.3.0-SNAPSHOT-320.jar:3.3.0-SNAPSHOT (git-08c03aae-b320)]
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[velocity-3.3.0-SNAPSHOT-320.jar:3.3.0-SNAPSHOT (git-08c03aae-b320)]
        at java.lang.Thread.run(Thread.java:1583) [?:?]
Caused by: java.net.ConnectException: connection timed out: sessionserver.mojang.com/13.107.246.74:443
        at org.asynchttpclient.netty.channel.NettyConnectListener.onFailure(NettyConnectListener.java:179) ~[velocity-3.3.0-SNAPSHOT-320.jar:3.3.0-SNAPSHOT (git-08c03aae-b320)]
        at org.asynchttpclient.netty.channel.NettyChannelConnector$1.onFailure(NettyChannelConnector.java:108) ~[velocity-3.3.0-SNAPSHOT-320.jar:3.3.0-SNAPSHOT (git-08c03aae-b320)]
        at org.asynchttpclient.netty.SimpleChannelFutureListener.operationComplete(SimpleChannelFutureListener.java:28) ~[velocity-3.3.0-SNAPSHOT-320.jar:3.3.0-SNAPSHOT (git-08c03aae-b320)]
        at org.asynchttpclient.netty.SimpleChannelFutureListener.operationComplete(SimpleChannelFutureListener.java:20) ~[velocity-3.3.0-SNAPSHOT-320.jar:3.3.0-SNAPSHOT (git-08c03aae-b320)]
        at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:590) ~[velocity-3.3.0-SNAPSHOT-320.jar:3.3.0-SNAPSHOT (git-08c03aae-b320)]
        at io.netty.util.concurrent.DefaultPromise.notifyListeners0(DefaultPromise.java:583) ~[velocity-3.3.0-SNAPSHOT-320.jar:3.3.0-SNAPSHOT (git-08c03aae-b320)]
        at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:559) ~[velocity-3.3.0-SNAPSHOT-320.jar:3.3.0-SNAPSHOT (git-08c03aae-b320)]
        at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:492) ~[velocity-3.3.0-SNAPSHOT-320.jar:3.3.0-SNAPSHOT (git-08c03aae-b320)]
        at io.netty.util.concurrent.DefaultPromise.setValue0(DefaultPromise.java:636) ~[velocity-3.3.0-SNAPSHOT-320.jar:3.3.0-SNAPSHOT (git-08c03aae-b320)]
        at io.netty.util.concurrent.DefaultPromise.setFailure0(DefaultPromise.java:629) ~[velocity-3.3.0-SNAPSHOT-320.jar:3.3.0-SNAPSHOT (git-08c03aae-b320)]
        at io.netty.util.concurrent.DefaultPromise.tryFailure(DefaultPromise.java:118) ~[velocity-3.3.0-SNAPSHOT-320.jar:3.3.0-SNAPSHOT (git-08c03aae-b320)]
        at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe$1.run(AbstractNioChannel.java:262) ~[velocity-3.3.0-SNAPSHOT-320.jar:3.3.0-SNAPSHOT (git-08c03aae-b320)]
        at io.netty.util.concurrent.PromiseTask.runTask(PromiseTask.java:98) ~[velocity-3.3.0-SNAPSHOT-320.jar:3.3.0-SNAPSHOT (git-08c03aae-b320)]
        at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:153) ~[velocity-3.3.0-SNAPSHOT-320.jar:3.3.0-SNAPSHOT (git-08c03aae-b320)]
        ... 8 more
Caused by: io.netty.channel.ConnectTimeoutException: connection timed out: sessionserver.mojang.com/13.107.246.74:443
        at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe$1.run(AbstractNioChannel.java:261) ~[velocity-3.3.0-SNAPSHOT-320.jar:3.3.0-SNAPSHOT (git-08c03aae-b320)]
        at io.netty.util.concurrent.PromiseTask.runTask(PromiseTask.java:98) ~[velocity-3.3.0-SNAPSHOT-320.jar:3.3.0-SNAPSHOT (git-08c03aae-b320)]
        at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:153) ~[velocity-3.3.0-SNAPSHOT-320.jar:3.3.0-SNAPSHOT (git-08c03aae-b320)]
        ... 8 more
```

--- a/proxy/src/main/java/com/velocitypowered/proxy/VelocityServer.java
+++ b/proxy/src/main/java/com/velocitypowered/proxy/VelocityServer.java
@@ -595,7 +595,7 @@ public class VelocityServer implements ProxyServer, ForwardingAudience {
   }
 
   public HttpClient getHttpClient() {
-    return null;
+    return httpClient;
   }
 
   public AsyncHttpClient getAsyncHttpClient() {

--- a/proxy/src/main/java/com/velocitypowered/proxy/VelocityServer.java
+++ b/proxy/src/main/java/com/velocitypowered/proxy/VelocityServer.java
@@ -71,6 +71,7 @@ import io.netty.channel.EventLoopGroup;
 import java.io.IOException;
 import java.io.InputStream;
 import java.net.InetSocketAddress;
+import java.net.http.HttpClient;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.security.AccessController;
@@ -143,6 +144,7 @@ public class VelocityServer implements ProxyServer, ForwardingAudience {
 
   private final ConnectionManager cm;
   private final ProxyOptions options;
+  private final HttpClient httpClient;
   private @MonotonicNonNull VelocityConfiguration configuration;
   private @MonotonicNonNull KeyPair serverKeyPair;
   private final ServerMap servers;
@@ -168,6 +170,7 @@ public class VelocityServer implements ProxyServer, ForwardingAudience {
     scheduler = new VelocityScheduler(pluginManager);
     console = new VelocityConsole(this);
     cm = new ConnectionManager(this);
+    httpClient = HttpClient.newHttpClient();
     servers = new ServerMap(this);
     serverListPingHandler = new ServerListPingHandler(this);
     this.options = options;
@@ -589,6 +592,10 @@ public class VelocityServer implements ProxyServer, ForwardingAudience {
   @Override
   public void closeListeners() {
     this.cm.closeEndpoints(false);
+  }
+
+  public HttpClient getHttpClient() {
+    return null;
   }
 
   public AsyncHttpClient getAsyncHttpClient() {

--- a/proxy/src/main/java/com/velocitypowered/proxy/connection/client/InitialLoginSessionHandler.java
+++ b/proxy/src/main/java/com/velocitypowered/proxy/connection/client/InitialLoginSessionHandler.java
@@ -217,6 +217,12 @@ public class InitialLoginSessionHandler implements MinecraftSessionHandler {
               return;
             }
 
+            if (throwable != null) {
+                logger.error("Unable to authenticate player", throwable);
+                inbound.disconnect(Component.translatable("multiplayer.disconnect.authservers_down"));
+                return;
+            }
+
             // Go ahead and enable encryption. Once the client sends EncryptionResponse, encryption
             // is enabled.
             try {

--- a/proxy/src/main/java/com/velocitypowered/proxy/connection/client/InitialLoginSessionHandler.java
+++ b/proxy/src/main/java/com/velocitypowered/proxy/connection/client/InitialLoginSessionHandler.java
@@ -218,9 +218,9 @@ public class InitialLoginSessionHandler implements MinecraftSessionHandler {
             }
 
             if (throwable != null) {
-                logger.error("Unable to authenticate player", throwable);
-                inbound.disconnect(Component.translatable("multiplayer.disconnect.authservers_down"));
-                return;
+              logger.error("Unable to authenticate player", throwable);
+              inbound.disconnect(Component.translatable("multiplayer.disconnect.authservers_down"));
+              return;
             }
 
             // Go ahead and enable encryption. Once the client sends EncryptionResponse, encryption

--- a/proxy/src/main/java/com/velocitypowered/proxy/connection/client/InitialLoginSessionHandler.java
+++ b/proxy/src/main/java/com/velocitypowered/proxy/connection/client/InitialLoginSessionHandler.java
@@ -42,19 +42,20 @@ import com.velocitypowered.proxy.protocol.packet.LoginPluginResponse;
 import com.velocitypowered.proxy.protocol.packet.ServerLogin;
 import io.netty.buffer.ByteBuf;
 import java.net.InetSocketAddress;
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
 import java.security.GeneralSecurityException;
 import java.security.KeyPair;
 import java.security.MessageDigest;
 import java.util.Arrays;
 import java.util.Optional;
-import java.util.concurrent.ExecutionException;
 import java.util.concurrent.ThreadLocalRandom;
 import net.kyori.adventure.text.Component;
 import net.kyori.adventure.text.format.NamedTextColor;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
-import org.asynchttpclient.ListenableFuture;
-import org.asynchttpclient.Response;
 import org.checkerframework.checker.nullness.qual.MonotonicNonNull;
 
 /**
@@ -77,7 +78,7 @@ public class InitialLoginSessionHandler implements MinecraftSessionHandler {
   private boolean forceKeyAuthentication;
 
   InitialLoginSessionHandler(VelocityServer server, MinecraftConnection mcConnection,
-      LoginInboundConnection inbound) {
+                             LoginInboundConnection inbound) {
     this.server = Preconditions.checkNotNull(server, "server");
     this.mcConnection = Preconditions.checkNotNull(mcConnection, "mcConnection");
     this.inbound = Preconditions.checkNotNull(inbound, "inbound");
@@ -209,63 +210,54 @@ public class InitialLoginSessionHandler implements MinecraftSessionHandler {
         url += "&ip=" + urlFormParameterEscaper().escape(playerIp);
       }
 
-      ListenableFuture<Response> hasJoinedResponse = server.getAsyncHttpClient().prepareGet(url)
-          .execute();
-      hasJoinedResponse.addListener(() -> {
-        if (mcConnection.isClosed()) {
-          // The player disconnected after we authenticated them.
-          return;
-        }
-
-        // Go ahead and enable encryption. Once the client sends EncryptionResponse, encryption
-        // is enabled.
-        try {
-          mcConnection.enableEncryption(decryptedSharedSecret);
-        } catch (GeneralSecurityException e) {
-          logger.error("Unable to enable encryption for connection", e);
-          // At this point, the connection is encrypted, but something's wrong on our side and
-          // we can't do anything about it.
-          mcConnection.close(true);
-          return;
-        }
-
-        try {
-          Response profileResponse = hasJoinedResponse.get();
-          if (profileResponse.getStatusCode() == 200) {
-            final GameProfile profile = GENERAL_GSON.fromJson(profileResponse.getResponseBody(),
-                GameProfile.class);
-            // Not so fast, now we verify the public key for 1.19.1+
-            if (inbound.getIdentifiedKey() != null
-                && inbound.getIdentifiedKey().getKeyRevision() == IdentifiedKey.Revision.LINKED_V2
-                && inbound.getIdentifiedKey() instanceof IdentifiedKeyImpl) {
-              IdentifiedKeyImpl key = (IdentifiedKeyImpl) inbound.getIdentifiedKey();
-              if (!key.internalAddHolder(profile.getId())) {
-                inbound.disconnect(
-                    Component.translatable("multiplayer.disconnect.invalid_public_key"));
-              }
+      HttpRequest httpRequest = HttpRequest.newBuilder().uri(URI.create(url)).build();
+      HttpClient.newHttpClient().sendAsync(httpRequest, HttpResponse.BodyHandlers.ofString())
+          .whenCompleteAsync((response, throwable) -> {
+            if (mcConnection.isClosed()) {
+              // The player disconnected after we authenticated them.
+              return;
             }
-            // All went well, initialize the session.
-            mcConnection.setActiveSessionHandler(StateRegistry.LOGIN,
-                new AuthSessionHandler(server, inbound, profile, true));
-          } else if (profileResponse.getStatusCode() == 204) {
-            // Apparently an offline-mode user logged onto this online-mode proxy.
-            inbound.disconnect(
-                Component.translatable("velocity.error.online-mode-only", NamedTextColor.RED));
-          } else {
-            // Something else went wrong
-            logger.error(
-                "Got an unexpected error code {} whilst contacting Mojang to log in {} ({})",
-                profileResponse.getStatusCode(), login.getUsername(), playerIp);
-            inbound.disconnect(Component.translatable("multiplayer.disconnect.authservers_down"));
-          }
-        } catch (ExecutionException e) {
-          logger.error("Unable to authenticate with Mojang", e);
-          inbound.disconnect(Component.translatable("multiplayer.disconnect.authservers_down"));
-        } catch (InterruptedException e) {
-          // not much we can do usefully
-          Thread.currentThread().interrupt();
-        }
-      }, mcConnection.eventLoop());
+
+            // Go ahead and enable encryption. Once the client sends EncryptionResponse, encryption
+            // is enabled.
+            try {
+              mcConnection.enableEncryption(decryptedSharedSecret);
+            } catch (GeneralSecurityException e) {
+              logger.error("Unable to enable encryption for connection", e);
+              // At this point, the connection is encrypted, but something's wrong on our side and
+              // we can't do anything about it.
+              mcConnection.close(true);
+              return;
+            }
+
+            if (response.statusCode() == 200) {
+              final GameProfile profile = GENERAL_GSON.fromJson(response.body(),
+                  GameProfile.class);
+              // Not so fast, now we verify the public key for 1.19.1+
+              if (inbound.getIdentifiedKey() != null
+                  && inbound.getIdentifiedKey().getKeyRevision() == IdentifiedKey.Revision.LINKED_V2
+                  && inbound.getIdentifiedKey() instanceof IdentifiedKeyImpl) {
+                IdentifiedKeyImpl key = (IdentifiedKeyImpl) inbound.getIdentifiedKey();
+                if (!key.internalAddHolder(profile.getId())) {
+                  inbound.disconnect(
+                      Component.translatable("multiplayer.disconnect.invalid_public_key"));
+                }
+              }
+              // All went well, initialize the session.
+              mcConnection.setActiveSessionHandler(StateRegistry.LOGIN,
+                  new AuthSessionHandler(server, inbound, profile, true));
+            } else if (response.statusCode() == 204) {
+              // Apparently an offline-mode user logged onto this online-mode proxy.
+              inbound.disconnect(
+                  Component.translatable("velocity.error.online-mode-only", NamedTextColor.RED));
+            } else {
+              // Something else went wrong
+              logger.error(
+                  "Got an unexpected error code {} whilst contacting Mojang to log in {} ({})",
+                  response.sslSession(), login.getUsername(), playerIp);
+              inbound.disconnect(Component.translatable("multiplayer.disconnect.authservers_down"));
+            }
+          }, mcConnection.eventLoop());
     } catch (GeneralSecurityException e) {
       logger.error("Unable to enable encryption", e);
       mcConnection.close(true);

--- a/proxy/src/main/java/com/velocitypowered/proxy/connection/client/InitialLoginSessionHandler.java
+++ b/proxy/src/main/java/com/velocitypowered/proxy/connection/client/InitialLoginSessionHandler.java
@@ -43,7 +43,6 @@ import com.velocitypowered.proxy.protocol.packet.ServerLogin;
 import io.netty.buffer.ByteBuf;
 import java.net.InetSocketAddress;
 import java.net.URI;
-import java.net.http.HttpClient;
 import java.net.http.HttpRequest;
 import java.net.http.HttpResponse;
 import java.security.GeneralSecurityException;
@@ -211,7 +210,7 @@ public class InitialLoginSessionHandler implements MinecraftSessionHandler {
       }
 
       HttpRequest httpRequest = HttpRequest.newBuilder().uri(URI.create(url)).build();
-      HttpClient.newHttpClient().sendAsync(httpRequest, HttpResponse.BodyHandlers.ofString())
+      server.getHttpClient().sendAsync(httpRequest, HttpResponse.BodyHandlers.ofString())
           .whenCompleteAsync((response, throwable) -> {
             if (mcConnection.isClosed()) {
               // The player disconnected after we authenticated them.

